lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,298 @@
+  llvm::DominatorTreeBase<Block, false> &domTree = domInfo.getDomTree(&funcReg);
+  // Get loop information to eliminate loop exits from the dependencies
+  CFGLoopInfo li(domTree);
+
+  DenseMap<Block *, SmallVector<Block *, 4>> control_deps_map;
+
+  // Initialize the control_deps_maps by creating an entry for every block
+  // constituting the region
+  for (Block &block : funcReg.getBlocks()) {
+    SmallVector<Block *, 4> deps;
+    control_deps_map.insert(std::make_pair(&block, deps));
+  }
+
+  for (Block &block : funcReg.getBlocks()) {
+    // Extract the dependencies of this block to adjust them by removing loop
+    // exit conditions, if any
+    // We adopt a convention of accessing the very last map of the vector
+    // because we only call this function at the end of
+    // identifyAllControlDeps() right after it adds a new entry to
+    // all_control_deps_maps
+    SmallVector<Block *, 4> block_deps =
+        all_control_deps_maps[all_control_deps_maps.size() - 1][&block];
+
+    for (auto &one_dep : block_deps) {
+      CFGLoop *loop = li.getLoopFor(one_dep);
+      if (loop == nullptr) {
+        // indicating that the one_dep is not inside any loop, so it must be a
+        // forward dependency
+        control_deps_map[&block].push_back(one_dep);
+      } else {
+        // indicating that the one_dep is inside a loop,
+        // to decide if it is a forward dep or not, compare it against all of
+        // the exits and latches of this loop
+        bool not_forward = false;
+        // check if one_dep is an exit of the loop
+        SmallVector<Block *> loop_exitBlocks;
+        loop->getExitingBlocks(loop_exitBlocks);
+        for (auto &loop_exit : loop_exitBlocks) {
+          if (loop_exit == one_dep) {
+            not_forward = true;
+            break; // it is not a forward dependency so no need to contiue
+                   // looping
+          }
+        }
+        // check if one_dep is a latch of the loop
+        SmallVector<Block *> loop_latchBlocks;
+        loop->getLoopLatches(loop_latchBlocks);
+        for (auto &loop_latch : loop_latchBlocks) {
+          if (loop_latch == one_dep) {
+            not_forward = true;
+            break; // it is not a forward dependency so no need to contiue
+                   // looping
+          }
+        }
+
+        if (!not_forward) {
+          control_deps_map[&block].push_back(one_dep);
+        }
+      }
+    }
+  }
+
+  forward_control_deps_maps.emplace_back(control_deps_map);
+}
+
+void ControlDependenceAnalysis::calculateBlockControlDeps(
+    Block *block, int funcOp_idx,
+    SmallVector<Block *, 4> &returned_control_deps) {
+  returned_control_deps = all_control_deps_maps[funcOp_idx][block];
+}
+
+void ControlDependenceAnalysis::calculateBlockForwardControlDeps(
+    Block *block, int funcOp_idx,
+    SmallVector<Block *, 4> &returned_forward_control_deps) {
+  returned_forward_control_deps = forward_control_deps_maps[funcOp_idx][block];
+}
```

Comment:
It is longer to write a call to these functions rather than to simply inline the code they contain at the call site, remove them. Their name is also confusing, they don't actually calculate anything.

Reply:
You can think of those functions as the getters of the class and those are the ones we use in the CfToHandshakePass to retrieve the control dependencies of a particular block (returned by reference through `returned_forward_control_deps`).

Reply:
Sounds good, then please start their name with "get" so that is is clear to the user what the purpose of these methods is. When I saw "calculate" my brain jumped immediately to "this is something used internally by the analysis to compute some of its state", hence my comment. I suspect this will be the case for other people as well.

There is also no need to start the name of the vector through which you return the data by reference by "return". When a non-const data-structure is passed by reference it usually indicates this very purpose. The method's documentation usually carries that meaning as well.